kotlin.html.markdown

```diff
@@ -0,0 +1,401 @@
+    // `obj` is automatically cast to `String` in this branch
+    return obj.length
+  }
+  
+  // else returns null
+}
+
+// Type checking and automatic casting, version2
+fun nullChecking1(obj: Any): Int? {
+  if (obj !is String)
+    return null
+    
+  // Automatic cast to `String`! we may use obj.length now.
+  return obj.length
+}
+
+// Automatic cast for right-hand side of `&&`
+fun nullChecking2(obj: Any): Int? {
+  if (obj is String && obj.length > 0)
+    return obj.length
+}
+
+// *If not null* shorthand 
+fun nullChecking3(obj: Any?): Int? {
+  println(obj?.length)
+}
+
+// *If not null* shorthand with else
+fun nullChecking4(obj: Any?): Int? {
+  println(obj?.size ?: "empty")
+}
+
+// The block inside let is executed only of obj is not null. No casting.
+fun nullChecking5(obj: Any): Int? {
+  object?.let {
+    return 42
+  }
+}
+
+// _______________________________________________________________ COLLECTIONS
+
+fun collectionIter() {
+  names = listOf("foo", "bar", "baz")
+  
+  // Iterating over collection items.
+  for (name in names)
+    println(name)
+
+  // Checking if collection contains an item.
+  // names.contains(text) is called automatically.
+  if (text in names) 
+    print("Yes")
+
+  // Function literals.
+  names filter { it.startsWith("A") } 
+    sortBy { it } 
+    map { it.toUpperCase() } 
+    forEach { print(it) }
+}
+
+// ___________________________________________________________________ CLASSES
+
+/**
+ * General class declaration syntax
+ *
+ * Note that *EVERYTHING* except class keyword and class name is optional!
+ *
+ * // @TODO other than private?
+ *
+ * class <NAME> [private] (PrimaryConstructorArgs) : someSuperType(ArgsIfAny) {
+ *   // init blocks, like so:
+ *   // secondary constructors.
+ *   // functions
+ *   // properties
+ *   // Nested and Inner Classes
+ *   // Object Declarations
+ * }
+ *
+ * Classes with no super type implicitly inherit from `Any`, but `Any` is not
+ * same as `java.lang.object`.
+ *
+ * [More about classes](http://kotlinlang.org/docs/reference/classes.html)
+ */
+
+// An empty class, no curly braces needed.
+class Empty
+
+/**
+ * A class with one primary constructors and more secondary constructors. A 
+ * primary constructor is defined inline.
+ * Both primary and secondary constructors are optional, but secondaries must 
+ * directly or indirectly (through other constructors) call the primary.
+ */
+public class LearnKotlin(firstName: String) {
+
+  // The code for primary constructor should be defined in init block. If it
+  // has any parameters, they will be available.
+  init {
+    logger.info("class initialized with value ${firstName}")
+  }
+  
+  // firstName from primary constructor available in initializer.
+  val myName = firstName.toUpperCase()
+  
+  // Secondary constructors are prefixed with *constructor*.
+  // Calling other constructors with *this* keyword.
+  constructor(val firstName: String, val lastName) this(firstName) {
```

Comment:
Only primary constructor parameters can be marked as 'var' or 'val'. Hint: edit the code part of your file in IntelliJ IDEA and make sure it doesn't highlight any errors or warnings.